# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [75]:
#Import your libraries
import pandas as pd
import numpy as np
import seaborn as sns



# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather` data. 

#### First, import it into a data frame called `austin`. 

In [76]:
# Your code here
austin = pd.read_csv('C:\\Users\\rapha\\lab-intro-to-ml\\your-code\\austin_weather.csv')
austin.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHighMiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       

[5 rows x 21 columns]

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [77]:
# Your code here
austin.describe(include='all')

Date    TempHighF     TempAvgF     TempLowF DewPointHighF  \
count         1319  1319.000000  1319.000000  1319.000000          1319   
unique        1319          NaN          NaN          NaN            64   
top     2016-02-12          NaN          NaN          NaN            74   
freq             1          NaN          NaN          NaN           112   
mean           NaN    80.862775    70.642911    59.902957           NaN   
std            NaN    14.766523    14.045904    14.190648           NaN   
min            NaN    32.000000    29.000000    19.000000           NaN   
25%            NaN    72.000000    62.000000    49.000000           NaN   
50%            NaN    83.000000    73.000000    63.000000           NaN   
75%            NaN    92.000000    83.000000    73.000000           NaN   
max            NaN   107.000000    93.000000    81.000000           NaN   

       DewPointAvgF DewPointLowF HumidityHighPercent HumidityAvgPercent  \
count          1319         1319                1319               1319   
unique           66           73                  58                 69   
top              71           66                 100                 64   
freq             84           62                 221                 51   
mean            NaN          NaN                 NaN                NaN   
std             NaN          NaN                 NaN                NaN   
min             NaN          NaN                 NaN                NaN   
25%             NaN          NaN                 NaN                NaN   
50%             NaN          NaN                 NaN                NaN   
75%             NaN          NaN                 NaN                NaN   
max             NaN          NaN                 NaN                NaN   

       HumidityLowPercent  ... SeaLevelPressureAvgInches  \
count                1319  ...                      1319   
unique                 82  ...                       101   
top                    52  ...                     29.95   
freq                   46  ...                        49   
mean                  NaN  ...                       NaN   
std                   NaN  ...                       NaN   
min                   NaN  ...                       NaN   
25%                   NaN  ...                       NaN   
50%                   NaN  ...                       NaN   
75%                   NaN  ...                       NaN   
max                   NaN  ...                       NaN   

       SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
count                       1319                1319               1319   
unique                       105                   5                 10   
top                        29.88                  10                 10   
freq                          48                1302                859   
mean                         NaN                 NaN                NaN   
std                          NaN                 NaN                NaN   
min                          NaN                 NaN                NaN   
25%                          NaN                 NaN                NaN   
50%                          NaN                 NaN                NaN   
75%                          NaN                 NaN                NaN   
max                          NaN                 NaN                NaN   

       VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
count                1319        1319       1319        1319   
unique                 12          22         13          37   
top                    10          13          4          20   
freq                  611         212        239         122   
mean                  NaN         NaN        NaN         NaN   
std                   NaN         NaN        NaN         NaN   
min                   NaN         NaN        NaN         NaN   
25%                   NaN         NaN        NaN         NaN   
50%           

In [ ]:
# Your code here


In [ ]:
# Your code here


#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [78]:
# Your code here
austin('DewPointAvgF').unique()

TypeError: 'DataFrame' object is not callable

In [79]:
# Your observation here
What do you think made Pandas to treat this column as object instead of int64? - # Text or Nan values

SyntaxError: invalid syntax (<ipython-input-79-ead11547d052>, line 2)

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [80]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [81]:
# Your code here
#austin.apply(wrong_type_columns.to_numeric, errors='coerce')
austin[wrong_type_columns] = austin[wrong_type_columns].apply(pd.to_numeric, errors='coerce')
austin.head(5)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  2013-12-21         74        60        45           67.0          49.0   
1  2013-12-22         56        48        39           43.0          36.0   
2  2013-12-23         58        45        32           31.0          27.0   
3  2013-12-24         61        46        31           36.0          28.0   
4  2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0  ...                      29.68                      29.59   
1  ...                      30.13                      29.87   
2  ...                      30.49                      30.41   
3  ...                      30.45                      30.30   
4  ...                      30.33                      30.27   

   VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  \
0                 10.0                 7.0                 2.0         20.0   
1                 10.0                10.0                 5.0         16.0   
2                 10.0                10.0                10.0          8.0   
3                 10.0                10.0                 7.0         12.0   
4                 10.0                10.0                 7.0         10.0   

   WindAvgMPH  WindGustMPH  PrecipitationSumInches               Events  
0         4.0         31.0                    0.46  Rain , Thunderstorm  
1         6.0         25.0                    0.00                       
2         3.0         12.0                    0.00                       
3         4.0         20.0                    0.00                       
4         2.0         16.0                     NaN                       

[5 rows x 21 columns]

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [82]:
# Your code here
austin.describe(include='all')

Date    TempHighF     TempAvgF     TempLowF  DewPointHighF  \
count         1319  1319.000000  1319.000000  1319.000000    1312.000000   
unique        1319          NaN          NaN          NaN            NaN   
top     2016-02-12          NaN          NaN          NaN            NaN   
freq             1          NaN          NaN          NaN            NaN   
mean           NaN    80.862775    70.642911    59.902957      61.515244   
std            NaN    14.766523    14.045904    14.190648      13.586452   
min            NaN    32.000000    29.000000    19.000000      13.000000   
25%            NaN    72.000000    62.000000    49.000000      53.000000   
50%            NaN    83.000000    73.000000    63.000000      66.000000   
75%            NaN    92.000000    83.000000    73.000000      73.000000   
max            NaN   107.000000    93.000000    81.000000      80.000000   

        DewPointAvgF  DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
count    1312.000000   1312.000000          1317.000000         1317.000000   
unique           NaN           NaN                  NaN                 NaN   
top              NaN           NaN                  NaN                 NaN   
freq             NaN           NaN                  NaN                 NaN   
mean       56.632622     50.933689            87.872437           66.671982   
std        14.866272     16.187480            11.038637           12.463634   
min         8.000000      2.000000            37.000000           27.000000   
25%        46.000000     38.000000            85.000000           59.000000   
50%        61.000000     56.000000            90.000000           67.000000   
75%        69.000000     65.000000            94.000000           74.000000   
max        76.000000     75.000000           100.000000           97.000000   

        HumidityLowPercent  ...  SeaLevelPressureAvgInches  \
count          1317.000000  ...                1316.000000   
unique                 NaN  ...                        NaN   
top                    NaN  ...                        NaN   
freq                   NaN  ...                        NaN   
mean             44.958998  ...                  30.022903   
std              16.974632  ...                   0.172145   
min              10.000000  ...                  29.550000   
25%              33.000000  ...                  29.910000   
50%              44.000000  ...                  30.000000   
75%              55.000000  ...                  30.100000   
max              93.000000  ...                  30.740000   

        SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
count                 1316.000000          1307.000000         1307.000000   
unique                        NaN                  NaN                 NaN   
top                           NaN                  NaN                 NaN   
freq                          NaN                  NaN                 NaN   
mean                    29.931892             9.991584            9.162204   
std                      0.172144             0.163489            1.458883   
min                     29.410000             5.000000            2.000000   
25%                     29.820000            10.000000            9.000000   
50%                     29.910000            10.000000           10.000000   
75%                     30.020000            10.000000           10.000000   
max                     30.610000            10.000000           10.000000   

        VisibilityLowMiles  WindHighMPH   WindAvgMPH  WindGustMPH  \
count          1307.000000  1317.000000  1317.000000  1315.000000   
unique                 NaN          NaN          NaN          NaN   
top                    NaN          NaN          NaN          NaN   
freq                   NaN          NaN          NaN          NaN   
mean              6.840092    13.231587     5.008352    21.373384   
std               3.683453     3.425096     2.086450    

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [83]:
# Your code here
missing_values = austin.isnull().any(axis=1)
missing_values.head(100)

0     False
1     False
2     False
3     False
4      True
      ...  
95     True
96    False
97    False
98    False
99    False
Length: 100, dtype: bool

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [84]:
# Your code here
austin.shape


(1319, 21)

In [85]:
missing_values.shape

(1319,)

#### Calculate the ratio of missing rows to total rows

In [86]:
# Your code here
missing_rows = austin.isnull().mean() * 100
missing_rows

Date                          0.000000
TempHighF                     0.000000
TempAvgF                      0.000000
TempLowF                      0.000000
DewPointHighF                 0.530705
DewPointAvgF                  0.530705
DewPointLowF                  0.530705
HumidityHighPercent           0.151630
HumidityAvgPercent            0.151630
HumidityLowPercent            0.151630
SeaLevelPressureHighInches    0.227445
SeaLevelPressureAvgInches     0.227445
SeaLevelPressureLowInches     0.227445
VisibilityHighMiles           0.909780
VisibilityAvgMiles            0.909780
VisibilityLowMiles            0.909780
WindHighMPH                   0.151630
WindAvgMPH                    0.151630
WindGustMPH                   0.303260
PrecipitationSumInches        9.401061
Events                        0.000000
dtype: float64

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [87]:
# Your code here
missing_rows2 = austin.isna().sum()
missing_rows2

Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
Events                          0
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [ ]:
# Your code here
# as showed before it was 9.401061

Almost 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [88]:
# Your code here 
austin.drop(axis=1, columns=['PrecipitationSumInches'], inplace=True)

# Print `austin` to confirm the column is indeed removed

austin

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH     

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [89]:
# Your code here
austin_fixed = austin.interpolate(inplace=False)
austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH     

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [90]:
# Your code here
missing_austin_fixed = austin_fixed.isna().sum()
missing_austin_fixed

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
Events                        0
dtype: int64

In [91]:
missing_austin = austin.isna().sum()
missing_austin

Date                           0
TempHighF                      0
TempAvgF                       0
TempLowF                       0
DewPointHighF                  7
DewPointAvgF                   7
DewPointLowF                   7
HumidityHighPercent            2
HumidityAvgPercent             2
HumidityLowPercent             2
SeaLevelPressureHighInches     3
SeaLevelPressureAvgInches      3
SeaLevelPressureLowInches      3
VisibilityHighMiles           12
VisibilityAvgMiles            12
VisibilityLowMiles            12
WindHighMPH                    2
WindAvgMPH                     2
WindGustMPH                    4
Events                         0
dtype: int64

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [92]:
# Your code here:
austin_fixed['Events'].value_counts()

                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: Events, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [ ]:
# Your answer:
Rain: 192

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [93]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here
for x in austin_fixed['Events']:
    if x == x in event_list:
        austin_fixed[x] = 1
    else:
        austin_fixed[x] = 0
    

# Print your new dataframe to check whether new columns have been created:

austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...               Events  Rain , Thunderstorm     \
0                   57.0  ...  Rain , Thunderstorm                    0  0   
1                   43.0  ...                                         0  0   
2                   27.0  ...                                         0  0   
3                   22.0  ...                                         0  0   
4                   56.0  ...                                         0  0   
...                  ...  ...                  ...                  ... ..   
1314                25.0  ...                                         0  0   
1315                20.0  ...                                         0  0   
1316                19.0  ...                                         0  0   
1317                27.0  ...                                         0  0   
1318                22.0  ...                                         0  0   

      Rain  Fog  Rain , Snow  Fog , Rain  Thunderstorm  \
0        1    1            0           0             1   
1        1    1            0           0             1   
2        1    1            0           0             1   
3        1    1            0           0             1   
4        1    1            0           0             1   
...    ...  ...          ...         ...           ...   
1314     1    1            0           0             1   
1315     1    1            0           0             1   
1316     1    1            0           0             1   
1317     1    1            0           0             1   
1318     1    1            0           0             1   

      Fog , Rain , Thunderstorm Fog , Thunderstorm  
0                             0                  0  
1                             0                  0  
2                             0                  0  
3                             0                  0  
4                             0                  0  
...                         ...                ...  
1314                          0                  0  
1315                          0                  0  
1316                          0                  0  
1317                          0                  0  
1318                          0                  0  

[1319 rows x 29 columns

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [ ]:
# Your code here
#already done with the for loop

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [ ]:
# Your code here
#already done above

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [94]:
# Your code here
austin_fixed.drop(axis=1, columns=['Events'], inplace=True)
austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  WindGustMPH  Rain , Thunderstorm     Rain  Fog  \
0                   57.0  ...         31.0                    0  0     1    1   
1                   43.0  ...         25.0                    0  0     1    1   
2                   27.0  ...         12.0                    0  0     1    1   
3                   22.0  ...         20.0                    0  0     1    1   
4                   56.0  ...         16.0                    0  0     1    1   
...                  ...  ...          ...                  ... ..   ...  ...   
1314                25.0  ...         21.0                    0  0     1    1   
1315                20.0  ...         20.0                    0  0     1    1   
1316                19.0  ...         17.0                    0  0     1    1   
1317                27.0  ...         20.0                    0  0     1    1   
1318                22.0  ...         20.0                    0  0     1    1   

      Rain , Snow  Fog , Rain  Thunderstorm  Fog , Rain , Thunderstorm  \
0               0           0             1                          0   
1               0           0             1                          0   
2               0           0             1                          0   
3               0           0             1                          0   
4               0           0             1                          0   
...           ...         ...           ...                        ...   
1314            0           0             1                          0   
1315            0           0             1                          0   
1316            0           0             1                          0   
1317            0           0             1                          0   
1318            0           0             1                          0   

      Fog , Thunderstorm  
0                      0  
1                      0  
2                      0  
3                      0  
4                      0  
...                  ...  
1314                   0  
1315                   0  
1316                   0  
1317                   0  
1318                   0  

[1319 rows x 28 columns]

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [95]:
# Your code here
from datetime import datetime as dt
austin_fixed(['Date']) = dt.dt.strptime('%Y-%m-%d').date()

#austin_fixed(['Date']).datetime.datetime.strptime

SyntaxError: can't assign to function call (<ipython-input-95-a2b6ed1a0d79>, line 3)

#### Print `austin_fixed` to check your `Date` column.

In [96]:
austin_fixed.head(5)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  2013-12-21         74        60        45           67.0          49.0   
1  2013-12-22         56        48        39           43.0          36.0   
2  2013-12-23         58        45        32           31.0          27.0   
3  2013-12-24         61        46        31           36.0          28.0   
4  2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  WindGustMPH  Rain , Thunderstorm     Rain  Fog  Rain , Snow  \
0  ...         31.0                    0  0     1    1            0   
1  ...         25.0                    0  0     1    1            0   
2  ...         12.0                    0  0     1    1            0   
3  ...         20.0                    0  0     1    1            0   
4  ...         16.0                    0  0     1    1            0   

   Fog , Rain  Thunderstorm  Fog , Rain , Thunderstorm  Fog , Thunderstorm  
0           0             1                          0                   0  
1           0             1                          0                   0  
2           0             1                          0                   0  
3           0             1                          0                   0  
4           0             1                          0                   0  

[5 rows x 28 columns]

In [104]:
sorted(austin_fixed)

[' ',
 'Date',
 'DewPointAvgF',
 'DewPointHighF',
 'DewPointLowF',
 'Fog',
 'Fog , Rain',
 'Fog , Rain , Thunderstorm',
 'Fog , Thunderstorm',
 'HumidityAvgPercent',
 'HumidityHighPercent',
 'HumidityLowPercent',
 'Rain',
 'Rain , Snow',
 'Rain , Thunderstorm',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureLowInches',
 'TempAvgF',
 'TempHighF',
 'TempLowF',
 'Thunderstorm',
 'VisibilityAvgMiles',
 'VisibilityHighMiles',
 'VisibilityLowMiles',
 'WindAvgMPH',
 'WindGustMPH',
 'WindHighMPH']

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [105]:
# Your code here:
y = austin_fixed['TempAvgF']
x = austin_fixed[[' ',
 'Date',
 'DewPointAvgF',
 'DewPointHighF',
 'DewPointLowF',
 'Fog',
 'Fog , Rain',
 'Fog , Rain , Thunderstorm',
 'Fog , Thunderstorm',
 'HumidityAvgPercent',
 'HumidityHighPercent',
 'HumidityLowPercent',
 'Rain',
 'Rain , Snow',
 'Rain , Thunderstorm',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureLowInches',
 'TempAvgF',
 'TempHighF',
 'TempLowF',
 'Thunderstorm',
 'VisibilityAvgMiles',
 'VisibilityHighMiles',
 'VisibilityLowMiles',
 'WindAvgMPH',
 'WindGustMPH',
 'WindHighMPH']]

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [106]:
#Your code here:
from sklearn.model_selection import train_test_split

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [107]:
#Your code here:
X_train, X_test, y_train, y_test = train_test_split(x, y,random_state = 0, test_size = 0.25)

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [ ]:
# Your code here:
